### PR TITLE
Fix Stack Monitoring with custom certificate without CA

### LINKED ISSUE #5309: Stack monitoring doesn't work with a resource monitored with a custom certificate without a CA
Stack Monitoring incorrectly assumes that if the monitored Elastic resource has TLS enabled, there is a CA to configure in the input section of the Metricbeat config. This overlloks that you can have a certificate issued by a well-known CA, so you don't always provide a CA when TLS is enabled. This makes Stack Monitoring impossible to use with a custom certificate without a CA.

Symptom: Metricbeat is in a boot loop due to an invalid configuration.
```
2022-01-31T23:09:16.106Z ERROR [kibana.stats] stats/stats.go:79 1 error: open 
/mnt/elastic-internal/kb-monitoring/<es-namespace>/<es-name>/certs/ca.crt: 
no such file or directory reading <nil> accessing config
```

### PR DESCRIPTION
**Pull request by thbkrkr**
Resolves #5309.

### CODE REVIEW

File: pkg/controller/common/certificates/ca_test.go [hunk 1]
```diff
@@ -85,3 +90,36 @@ func TestNewSelfSignedCA(t *testing.T) {
...
+			CertFileName: []byte("..."),
+		},
+	}
+	hasCA, err = HasPublicCA(k8s.NewFakeClient(&certsWithoutCA), esv1.ESNamer, "ns", "c2")
+	assert.NoError(t, err)
+	assert.False(t, hasCA)
+
+	// no certs secret
+	_, err = HasPublicCA(k8s.NewFakeClient(), esv1.ESNamer, "ns", "c3")
+	assert.Error(t, err)
```
Comment (pebrc) [inline pkg/controller/common/certificates/ca_test.go:124]: Should we stick with the table tests for consistency?

File: pkg/controller/common/stackmon/config.go [hunk 2]
```diff
@@ -159,8 +159,8 @@ type inputConfigData struct {
 	URL      string
 	Username string
 	Password string
-	IsSSL    bool
-	SSLPath  string
+	IsCA     bool
```
Comment (pebrc) [inline pkg/controller/common/stackmon/config.go:162]: I know `Is*` is the go convention but would `HasCA` not be better fitting here?

File: pkg/controller/kibana/stackmon/metricbeat.tpl.yml [hunk 3]
```diff
@@ -9,8 +9,8 @@ metricbeat.modules:
     hosts: ["{{ .URL }}"]
     username: {{ .Username }}
     password: {{ .Password }}
-    {{- if .IsSSL }}
-    ssl.certificate_authorities: ["{{ .SSLPath }}"]
+    {{- if .IsCA }}
+    ssl.certificate_authorities: ["{{ .CAPath }}"]
     ssl.verification_mode: "{{ .SSLMode }}"
```
Comment (pebrc) [inline pkg/controller/kibana/stackmon/metricbeat.tpl.yml:14]: `verification_mode` is now predicated on the presence of a CA certificate. Will this not be a problem if users are using their custom certificates without a CA and those certificates are not issued for localhost?

File: pkg/controller/common/certificates/ca.go [hunk 4]
```diff
@@ -124,3 +130,18 @@ func (c *CA) CreateCertificate(
 
 	return certData, err
 }
+
+// HasPublicCA returns true if an Elastic resource has a CA (ca.crt) in its public HTTP certs secret given its namer,
+// namespace and name.
+func HasPublicCA(client k8s.Client, namer name.Namer, namespace string, name string) (bool, error) {
```
Comment (barkbay) [inline pkg/controller/common/certificates/ca.go:136]: nit, naming: `ca.crt` can be managed in the context of a private PKI, without being a public CA

```suggestion
// PublicCertsHasCACert returns true if an Elastic resource has a Certificate Authority certificate (ca.crt) in its public HTTP certs secret given its namer,
// namespace and name.
func PublicCertsHasCACert(client k8s.Client, namer name.Namer, namespace string, name string) (bool, error) {
```

**Review by barkbay — COMMENTED** [review]
I did a quick test with Elasticsearch and Kibana pub. certs signed by a well known CA. Metricbeat is failing with the following error:

```
2022-02-01T08:26:57.702Z	ERROR	module/wrapper.go:259	Error fetching data for metricset elasticsearch.index: error determining if connected Elasticsearch node is master: error making http request: Get "https://localhost:9200/_nodes/_local/nodes": x509: certificate is valid for my-public-hostname.com, not localhost
```

I think it is related to Peter's comment [here](https://github.com/elastic/cloud-on-k8s/pull/5310#discussion_r796321746).

File: pkg/controller/common/certificates/ca_test.go (hunk 1 above)
Comment (thbkrkr) [reply pkg/controller/common/certificates/ca_test.go:124]: Yes.

File: pkg/controller/common/stackmon/config.go (hunk 2 above)
Comment (thbkrkr) [reply pkg/controller/common/stackmon/config.go:162]: ++, I was tempted but I didn't dare.

File: pkg/controller/kibana/stackmon/metricbeat.tpl.yml (hunk 3 above)
Comment (thbkrkr) [reply pkg/controller/kibana/stackmon/metricbeat.tpl.yml:14]: It is set to `certificate` which IIUC should verify that the provided certificate is signed by a trusted authority (CA), but does not perform any hostname verification ([doc](https://www.elastic.co/guide/en/beats/metricbeat/current/configuration-ssl.html#ssl-common-config)). But you are right, this is a problem as reported by Michael [here](https://github.com/elastic/cloud-on-k8s/pull/5310#pullrequestreview-868801747). I'm a bit confused.

Comment (thbkrkr) [reply pkg/controller/kibana/stackmon/metricbeat.tpl.yml:14]: Yes, you are right, it also needs to be set with a custom cert without a CA.

Comment (thbkrkr) [reply pkg/controller/kibana/stackmon/metricbeat.tpl.yml]: I opted to keep `IsSSL`, add `ssl.enabled: {{ .IsSSL }}` and always set `ssl.verification_mode: certificate`.

**Comment by thbkrkr** [comment]
Thanks for the feedback. All have been addressed.

File: pkg/controller/elasticsearch/stackmon/metricbeat.tpl.yml [hunk 5]
```diff
@@ -17,9 +17,10 @@ metricbeat.modules:
     hosts: ["{{ .URL }}"]
     username: {{ .Username }}
     password: {{ .Password }}
-    {{- if .IsSSL }}
-    ssl.certificate_authorities: ["{{ .SSLPath }}"]
+    ssl.enabled: {{ .IsSSL }}
     ssl.verification_mode: "{{ .SSLMode }}"
```
Comment (pebrc) [inline pkg/controller/elasticsearch/stackmon/metricbeat.tpl.yml:21]: Nit: I wonder if we should just have the verification mode in the template directly given that the value is effectively constant.

**Review by pebrc — APPROVED** [review]
Looks good to me now. I did a test with a Let's encrypt certificate and Elasticsearch and one with the self-signed certs ECK uses by default.

File: pkg/controller/elasticsearch/stackmon/metricbeat.tpl.yml (hunk 5 above)
Comment (thbkrkr) [reply pkg/controller/elasticsearch/stackmon/metricbeat.tpl.yml]: Yeah, that crossed my mind too.

**Comment by thbkrkr** [comment]
I found a little bug. When disabling TLS, we generate certs for ES but not for KB. This means we shouldn't get the `kb-http-public-certs` secret to resolve `hasCA` for the Kibana Metricbeat config without first checking if TLS is enabled.

**Review by pebrc — APPROVED** [review]
> I found a little bug

Good catch. I did not test with Kibana initially, but it looks good now.

**Pull request merged**